src/rgen/Microsoft.Macios.Bindings.Analyzer/NativeObjectHandleAnalyzer.cs

```diff
@@ -0,0 +1,166 @@
+			return methodSymbol.ContainingType;
+		return methodSymbol.ReturnType;
+	}
+
+	private void AnalyzeNode (SyntaxNodeAnalysisContext context)
+	{
+		if (context.Node is not MemberAccessExpressionSyntax memberAccess)
+			return;
+
+		if (memberAccess.Name.Identifier.Text != "Handle" &&
+			memberAccess.Name.Identifier.Text != "GetHandle" &&
+			memberAccess.Name.Identifier.Text != "GetNonNullHandle" &&
+			memberAccess.Name.Identifier.Text != "GetCheckedHandle")
+			return;
+
+		var symbol = context.SemanticModel.GetSymbolInfo (memberAccess.Expression).Symbol;
+		if (symbol is null)
+			return;
+
+		if ((symbol as IParameterSymbol)?.IsThis == true)
+			return;
+
+		if (symbol is IMethodSymbol methodSymbol) {
+			if (IsINativeObject (GetRealReturnType (methodSymbol)) &&
+				!IsSafeMethod (methodSymbol)) {
+				// Calling Handle directly on a value returned from method is wrong because
+				// the object is immediately collectible.                
+				var diagnostic = Diagnostic.Create (RBI0014, memberAccess.GetLocation (), symbol.Name);
+				context.ReportDiagnostic (diagnostic);
+			}
+			return;
+		}
+
+		var type = (symbol as ILocalSymbol)?.Type ?? (symbol as IParameterSymbol)?.Type;
+		if (type is null || !IsINativeObject (type))
+			return;
+
+		// Ignore variables that are wrapped in using block
+		var usingStatement = memberAccess.Parent?.FirstAncestorOrSelf<UsingStatementSyntax> ();
+		while (usingStatement is not null) {
+			if (usingStatement.Expression is not null) {
+				var resourceSymbol = context.SemanticModel.GetSymbolInfo (usingStatement.Expression).Symbol;
+				if (resourceSymbol is not null && SymbolEqualityComparer.Default.Equals (resourceSymbol, symbol))
+					return;
+			}
+			if (usingStatement.Declaration is not null) {
+				foreach (var variable in usingStatement.Declaration.Variables) {
+					var declaredSymbol = context.SemanticModel.GetDeclaredSymbol (variable);
+					if (declaredSymbol is not null && SymbolEqualityComparer.Default.Equals (declaredSymbol, symbol))
+						return;
+				}
+			}
+			usingStatement = usingStatement.Parent?.FirstAncestorOrSelf<UsingStatementSyntax> ();
+		}
+
+		bool accessedAfter = false;
+		var block = memberAccess.FirstAncestorOrSelf<BlockSyntax> ();
+		if (block is null) {
+			// We are not in a block. This can happen when handles are accessed in calls in constructor
+			// parameters, or if we are in an expression body. For constructors we have to check the
+			// method body for GC.KeepAlive or other references.
+			var constructor = memberAccess.FirstAncestorOrSelf<ConstructorDeclarationSyntax> ();
+			if (constructor is not null &&
+				constructor.Body?.Statements is { } statements &&
+				statements.Count > 0) {
+				var df = context.SemanticModel.AnalyzeDataFlow (statements.First (), statements.Last ());
+				if (df is null)
+					return;
+				accessedAfter = df.ReadInside.Contains (symbol);
+			} else {
+				// Assume expressions don't access the variable
+			}
```

Comment:
Empty else clause, what do we want to do here?

Reply:
It's a branch taken when constructor calls `: this(...)` or `: base(...)`, accesses `.Handle` in one of the parameters and the method body of the constructor is missing or an expression.

In this case `accessedAfter` is already set to `false` and it will produce the diagnostic error.

Reply:
(I can expand the comment to explain it)